Baskerville/llama3-inference-srun.md

```diff
@@ -0,0 +1,254 @@
+# Llama 3 inference on Baskerville using `srun`
+
+Before you can make use of them you’ll need to download the models.
+See [llama-download.md](llama-download.md) for how to go about doing this on Baskerville.
+
+# Llama 3 8B inference, single node, single GPU
+
+The 8B parameter Llama 3 model will run with a single process on a single node with a single GPU.
+The checkpoint provided for use with the example will, in fact, only work with a single process.
+
+We found at least 32 GiB of RAM was needed to run the model.
+With too little RAM the process will be sent a SIGKILL termination signal, the process will end with an exit code of -9 and an OOM report will be generated.
+
+The RAM, node, GPU and process quantities must be explicitly configured using `srun` for things to work.
+
+## 1. Log in to a single node with a single GPU and a suitable configuration:
+
+You’ll need to replace `<PROJECT_ID>` with the name of your project.
+
+```shell
+$ srun --qos turing \
+    --account <PROJECT_ID> \
+    --time 0:30:0 \
+    --nodes 1 \
+    --gpus 1 \
+    --mem 32768 \
+    --pty bash
+```
+
+## 2. Set up the environment on the node:
+
+```shell
+$ cd llama3
+$ module purge
+$ module load baskerville
+$ module load bask-apps/live
+$ module load PyTorch/2.1.2-foss-2022b-CUDA-11.8.0
+$ python -m venv venv
+$ . ./venv/bin/activate
+$ pip install pip --upgrade
+$ pip install -e .
+$ pip install -r requirements.txt
+```
+
+## 3. Run the example script that performs a bunch of inference steps:
+
+```shell
+$ python -m torch.distributed.run \
+    --standalone \
+    --nnodes 1 \
+    --nproc_per_node 1 \
+    example_chat_completion.py \
+    --ckpt_dir Meta-Llama-3-8B-Instruct/ \
+    --tokenizer_path Meta-Llama-3-8B-Instruct/tokenizer.model \
+    --max_seq_len 512 \
+    --max_batch_size 6
+```
+
+You should see some output that looks something like this:
+
+```text
+[...]
+Loaded in 10.39 seconds
+[...]
+==================================
+
+System: Always answer with Haiku
+
+User: I am going to Paris, what should I see?
+
+> Assistant: Eiffel's iron kiss
+River Seine's gentle whispers
+Art in every stone
+
+==================================
+```
+
+One things to note is that in the Llama 3 docs it [recommends](https://github.com/meta-llama/llama3?tab=readme-ov-file#quick-start) to use `torchrun` to execute the script.
+When `torchrun` is used directly in thisß way it doesn’t support Python virtual environments because it defaults to the system-installed Python.
```

Comment:
```suggestion
When `torchrun` is used directly in this way it doesn’t support Python virtual environments because it defaults to the system-installed Python.
```